Update the Spinner design

jameskoster opened the pull request:
This PR updates the Spinner component based on the designs in #36725. It also adjusts/fixes the position of the Spinner that appears when searching for documents in the add-link interface. Finally, it adds the gray-300 value to color-values.js, which has until now been missing. On this point I did notice that gray-200 has the incorrect value, but as I'm not sure where that is used, I figured we should fix it separately.

Things seem to be working well in the following scenarios:

* Loading a document that has internal dependencies like template parts, or the site logo
* Embedding third party content like a Tweet.
* Searching blocks in the Inserter
* Searching in the link input

<img src="https://user-images.githubusercontent.com/846565/149160764-8e0cbc4c-79eb-423c-bc7c-ec5f622fe993.gif" width="600" />


### Testing instructions

* Perform any action that invokes a spinner and check that it displays and behaves as expected, examples:
  * Edit a template in the site editor that includes template parts
  * Embed third party content like a tweet
  * Search for a block in the Inserter
  * Add a link and search for a post/page
  * Others that I'm missing? :)
* Try different browsers. This is working in Chrome, Firefox, and Safari, but I am unable to test Edge

jameskoster committed:
Update the spinner

jameskoster committed:
Formatting

jameskoster committed:
Spinner position in link input

jameskoster committed:
Use a span instead of a progress

jameskoster committed:
formatting

jameskoster committed:
Positioning

jasmussen commented:
This is very nice! Thanks for working on it.

I also really love these designs:
<img width="613" alt="Screenshot 2022-01-10 at 09 48 00" src="https://user-images.githubusercontent.com/1204802/148738922-3c3f0068-e8db-495a-9e31-0cc7db7dc774.png">

Is there any way we can get closer to these in terms of design? Stroke-width, white instead of spot color in dark contexts, a gray base circle track?

jameskoster committed:
Try an svg instead

jameskoster commented:
The only way I can think to include the track would be to use an svg rather than css alone, otherwise the mask is interfering too much. I've pushed an update to try this. 


https://user-images.githubusercontent.com/846565/148800185-1bc87154-5c99-447e-a1cd-811793ad959e.mp4



I'm not sure what the stroke width is in those screenshots, but I've set them to 1.5px for now. Easily changed.

I can't think of a way to make the colors adapt based on the background as the Spinner isn't aware of that. This might be something better left as an option on the Spinner itself.

jasmussen commented on packages/components/src/spinner/style.scss:
```diff
@@ -0,0 +1,44 @@
+.components-spinner--svg {
```
The SVG is working fairly well here! However just as far as classnames go, should we call it "track" instead? Also, I think it should be __ instad of --. I.e. `.components-spinner__track` for example.

jasmussen commented on packages/components/src/spinner/style.scss:
```diff
@@ -0,0 +1,44 @@
+	left: 0;
+}
+
+@keyframes rotatesvg {
```
Same comment with above, denoting the element as "track" might be better.

It seems correct to me to put this keyframe statement right in this components file where it will be used, especially since it's unique to this component and not reused. But naming-wise, we could use some BEM niceness here as well, for example `components-spinner__spin-track` perhaps?

jasmussen commented on packages/components/src/spinner/style.scss:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+.components-spinner--circle {
```
If I understand correctly, this is the blue part of the spinner that in this PR contracts and expands, so it's more of a quarter-circle. I realize this is nitpick territory, but I think it'd help readability. How about something like `.components-spinner__indicator`? Note the __ instead of --.

jasmussen commented on packages/components/src/spinner/style.scss:
```diff
@@ -0,0 +1,44 @@
+	transform-origin: 50% 50%;
+}
+
+@keyframes rotatecircle {
```
BEM all the way. How about `components-spinner__spin-indicator`?

jasmussen reviewed (commented):
The SVG is working really well here. For testing it I managed to record this GIF:
![spin](https://user-images.githubusercontent.com/1204802/148940381-54392e8d-93ab-4139-8b21-a94810e29739.gif)

I can't help but feel like the material influence here makes it a bit stuttery. The way the expansion and contraction of the blue indicator makes it appear as if it's pausing. I may be rather dull in the following, but a simple uniform rotation with no contracting might actually be preferable.

I'm going to tinker a bit in this codepen and come back! Thanks again for the work 👌

jasmussen commented:
Another reason to maybe reconsider the expansion and contraction of the indicator is that it isn't relative to the size of the spinner. While I don't know exactly how often we'll show a 128x128 spinner, a few things broke apart when I tried it:

![spin 128](https://user-images.githubusercontent.com/1204802/148941544-af79b018-b876-40b0-bfa0-c10bd2373410.gif)

It's actually still expanding and contracting, it's just not very visible because the dashoffset change is insignificant when large. I wonder if you could use `em`s there? Probably not? 🤔 

Another one is the fact that the gray track is outset compared to the blue indicator. I don't know why that happens. **Edit: Oh I know, the radius of the circle is 8.25, it should be 9 to match with half of the 18x18 svg viewbox. Combined with overflow: visible; on the SVG, it works.**

jasmussen commented:
Using an SVG seems appropriate enough for this, but I'm starting to think that using `dasharray` might not be. Mainly because I can't figure out how to make a dasharray that behaves in a relatively scaling way. Shown here a dasharray that approaches (but isn't precisely) a quarter circle as from the initial mockup, but only when the spinner is really big:
<img width="221" alt="Screenshot 2022-01-11 at 13 41 35" src="https://user-images.githubusercontent.com/1204802/148944664-54601ec7-33f6-40e5-9e9e-5e3d8c5d2569.png">

We might need a static SVG that just draws that quarter-circle precisely, and not use the dasharray trick 🤔 

Edit: Feel free to tinker in this codepen: https://codepen.io/joen/pen/RwLeByP?editors=1100

jameskoster commented:
> because I can't figure out how to make a dasharray that behaves in a relatively scaling way

Perhaps we can add some logic so that the dasharray value(s) are based on the spinner size?

I appreciate it's subjective/intangible, but I quite like the expansion/contraction effect. I can't put my finger on why but it makes it _feel_ like something is happening compared with the linear spinner. Maybe it's the pulsing.

Then again, if we eliminate that effect it would make the original design much easier to accomplish without the need for an svg. Maybe we can add a pulsing effect another way 🤔 I will also have a play in codepen :D

jameskoster commented:
I added a "Pulsey no svg" version to [this codepen](https://cdpn.io/jameskoster/debug/OJxBrGE).

Gif:

![spinner](https://user-images.githubusercontent.com/846565/148970557-e99cbf5f-8a7b-49f5-b908-cb4f605858ac.gif)

The main drawback to this approach is that it uses borders for the indicator line, which cannot be set to a 1.5px width.

jasmussen commented:
Impressive work going on in these codepens, I love it. I'm also starting to see your point about the contraction/expansions being a bit more dynamic. But just looking at them side by side, I'm personally still most attracted to the complete calmness of the linearly rotating circle, I feel like it might just work the best in noisy contexts.

jameskoster committed:
Revert to html + css, and simplify the animation

jameskoster commented:
Reverted to html + css only, and simplified the animation. Here's the result:

![spinner](https://user-images.githubusercontent.com/846565/148979668-6b8bb4f2-509a-46ca-8d83-f392707e1129.gif)

The last question seems to be what to do about the discrepancy between track/indicator width.

jasmussen commented:
Clever trick to use border-top to grab exactly one quarter of the circle! I suppose box-shadow wouldn't work for that, as it would be affected by the contours of the circle shape. I do think it's important we get the right stroke width, though — maybe the SVG was fine after all? That might also let us keep that neat rounded edgecap.

Alternatively we could create another full circle and mask a quadrant of it?

jameskoster committed:
Back to SVG

jameskoster commented:
To get the strokes to be 1.5px I think we'll have to return to the SVG approach. I just pushed a change that does this, and includes logic so that the dasharray will scale correctly. You can change the `spinnerSize` value in config-values.js to test this. Here's an example at 180px:

<img src="https://user-images.githubusercontent.com/846565/149142250-230f238d-6501-452e-9876-e452c5b1db42.gif" width="350" />

jasmussen commented:
I mean, I really dig that one 👌 

Is this one ready for final review?

jameskoster commented:
Yup, I think so!

jameskoster committed:
gray 400 -> 300

jasmussen commented on packages/block-editor/src/components/link-control/style.scss:
```diff
@@ -482,16 +482,11 @@ $preview-image-height: 140px;
-		top: $grid-unit-20 + 1px + ( ( 40px - $spinner-size ) * 0.5 );
-		right: $grid-unit-20 + 1px + ( $button-size * $block-editor-link-control-number-of-actions ) + ( ( $button-size - $spinner-size ) * 0.5 );
+		top: $grid-unit-15;
+		right: $button-size;
```
Glorious simplification, very nice.

jasmussen commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,52 @@
+	width: ${ CONFIG.spinnerSize }px;
+	height: ${ CONFIG.spinnerSize }px;
+	box-shadow: inset 0 0 0 0.75px ${ COLORS.gray[ 300 ] },
+		0 0 0 0.75px ${ COLORS.gray[ 300 ] };
```
Should we add a comment describing why there are two shadows stacked? Something like "The SVG can only have a centered stroke, and a stacked box-shadow accomplishes a matching 1.5px border.

jasmussen commented on packages/components/src/utils/colors-values.js:
```diff
@@ -31,6 +31,7 @@ export const G2 = {
 		600: '#949494', // Meets 3:1 UI or large text contrast against white.
 		400: '#ccc',
+		300: '#ddd',
 		200: '#ddd', // Used for most borders.
```
Hold up, now 200 and 300 are the same 🤔 

I guess the error is that 200 should be `#e0e0e0`:

<img width="562" alt="Screenshot 2022-01-13 at 09 24 29" src="https://user-images.githubusercontent.com/1204802/149293065-ed90b182-5116-4f23-b10b-ab35477c0901.png">

Since it looks like 200 was already erroneous before this PR, it seems fine to not fix that in this one, but this is an error we should correct.

jameskoster replied:
Yeah I noted that on the updated OP and agree, we should fix it separately.

ciampo replied:
Opened https://github.com/WordPress/gutenberg/pull/38109 to address this separately

ciampo replied:
#38109 has been merged, this PR will likely require a rebase to solve conflicts

jasmussen commented on packages/components/src/utils/config-values.js:
```diff
@@ -48,7 +48,7 @@ export default {
 	borderWidthFocus: '1.5px',
 	borderWidthTab: '4px',
-	spinnerSize: '18px',
+	spinnerSize: '18',
```
Just for my own curiosity, why is it unitless?

jameskoster replied:
In order for the spinner to scale correctly this value is used on attributes like the svg viewbox, and the circle radius. I think the radius would accept a unit, but unfortunately the viewbox does not.

ciampo replied:
Since it's unitless, should be change it to be a number instead of a string?

ciampo replied:
Marking as resolved, in favour of https://github.com/WordPress/gutenberg/pull/37551#discussion_r788614818

jasmussen reviewed (approved):
This is working really well for me:
![spin](https://user-images.githubusercontent.com/1204802/149293267-662d70aa-0bb7-421b-b704-5509c98a1364.gif)

It's a nice and simple spinner that benefits from the spot color. 

I know you had your heart set on some expansion/contraction of the indicator, I think that's still something that can be explored in a codepen if you're finding yourself missing it. But because it's so rarely shown (thankfully things load fast), starting minimalist like this feels like a good baseline.

I left a few comments, but this looks good. Nice work, thank you!

jameskoster committed:
Add a comment explaining the wrapper and the shadow

sarayourfriend commented on packages/components/src/spinner/style.scss:
```diff
@@ -0,0 +1,32 @@
+.components-spinner {
```
Sorry to pop in 🙂 Why does this component use a mix of Scss and Emotion? Is it possible for it to just use Emotion? (I think that's the way things are headed)

Also, if this component replaces the existing spinner, we could probably go ahead and remove the spinner in `components/src/ui/spinner` as well to avoid any confusion for future contributors to the components package.

cc @ciampo

jameskoster replied:
> Sorry to pop in 🙂

Welcome! :D I'm using emotion where I need to inject dynamic values (like the spinner size) in to the css, then scss everywhere else. I figured plain scss is more contributor friendly. 

It certainly could be all-emotion if that's preferable.

> Also, if this component replaces the existing spinner, we could probably go ahead and remove the spinner in components/src/ui/spinner as well to avoid any confusion for future contributors to the components package.

Do you think it would be ok to do that in a follow-up? I'm not entirely sure how/where it is used.

sarayourfriend replied:
> It certainly could be all-emotion if that's preferable.

I think @ciampo should be able to answer that. I'm not familiar with the intentions of the components package since I stopped working on it a few months ago. If he doesn't get back to you in a timely manner I'd say a fast-follow PR to convert it to all Emotion would probably be fine.

> Do you think it would be ok to do that in a follow-up? I'm not entirely sure how/where it is used.

Sure thing 👍 FWIW I think it is unused so will hopefully be a straightforward PR (but don't quote me on that! 😅)

jameskoster replied:
I think Woo may have used it, so to be safe I'll do that separately :)

sarayourfriend replied:
It doesn't look like the `ui/spinner` is exported from the components package, so hopefully Woo isn't doing something to force it to be available 😱

jameskoster replied:
Oh, I just realised that ui/spinner was the original intended replacement for the Spinner that I worked on in this PR (IE that one that is actually published and in use). In which case it would seem very safe to remove ui/spinner. I'll still probably do that separately, though :D

ciampo replied:
Thank you @sarayourfriend for the ping — I haven't taken a deeper look yet, but we would definitely like to [use Emotion to style our components](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#styling).

@jameskoster , would you be able to move back to CSS-in-JS styles (Emotion) in this PR?

The removal of `ui/spinner` can definitely happen in a separate PR

jameskoster replied:
> would you be able to move back to CSS-in-JS styles (Emotion) in this PR?

Yup np, I just pushed an update.

ciampo replied:
> Yup np, I just pushed an update.

Thank you! One last thing on this end — would you mind moving the code for the styles (i.e. `spinAnimation`, `StyledSpinner` and `SpinnerIndicator`) to a separate file? `packages/components/src/spinner/styles.js` would be perfect

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+			<SVG
+				viewBox={ viewBox }
+				xmlns="http://www.w3.org/2000/svg"
+				className="components-spinner__track"
```
After converting to Emotion, let's remove this hardcoded classname

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+					cx="50%"
+					cy="50%"
+					r={ radius }
+					stroke-dasharray={ strokeDasharray }
```
I believe in React this prop should be specific camel-case, i.e. `strokeDasharray`

ciampo replied:
This was actually done in https://github.com/WordPress/gutenberg/pull/37551/commits/cd30bdb1fb1c1de37ea37953edf1305425300385, marking as resolved

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+				focusable="false"
+			>
+				<Circle
+					className="components-spinner__indicator"
```
After converting to Emotion, let's remove this hardcoded classname

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+					cy="50%"
+					r={ radius }
+					stroke-dasharray={ strokeDasharray }
+					vector-effect="non-scaling-stroke"
```
I believe in React this prop should be specific camel-case, i.e. `vectorEffect`

ciampo replied:
Done in https://github.com/WordPress/gutenberg/pull/37551/commits/cd30bdb1fb1c1de37ea37953edf1305425300385, marking as resolved

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+				className="components-spinner__track"
+				role="img"
+				aria-hidden="true"
+				focusable="false"
```
Why is this attribute necessary? SVG elements shouldn't be focusable by default

jasmussen replied:
I'm fairily sure we've been adding `focusable="false"` because IE11 would focus SVGs regardless. Since we don't support IE11 anymore, we can remove it, however I wonder if we shouldn't do that in a separate PR, and remove across all SVGs in the codebase? Not a strong opinion.

ciampo replied:
Thank you for the additional context!

> I wonder if we shouldn't do that in a separate PR, and remove across all SVGs in the codebase?

Makes sense to me. Let's tackle this separately

jameskoster replied:
Similar to my comment on a11y, let's keep this PR focussed on the visual stuff and address these details separately?

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+				viewBox={ viewBox }
+				xmlns="http://www.w3.org/2000/svg"
+				className="components-spinner__track"
+				role="img"
+				aria-hidden="true"
```
Using `role="img"` and `aria-hidden="true"` seems contradictory: on one hand we're telling the browser that this is a meaningful image (for which we don't provide alt text), on the other we're telling the browser (and any assistive technology) to ignore this element.

I'm not an expert in a11y patterns, but from a quick look, it seems like a good combination could be using `role="alert"` and `aria-live="assertive"` ([example from carbon design system](https://www.carbondesignsystem.com/components/loading/accessibility)). If we don't want to investigate this route too much, in my opinion we should at least:

- remove `aria-hidden="true"`
- keep the `role="img"` attribute
- provide an accessible label. According to [this article](https://css-tricks.com/accessible-svgs/), the best way to achieve this would be to add a `<title id="uniqueTitleID">The title of the SVG</title>` tag inside the SVG element, and then to add `aria-labelledby="uniqueTitleID"` to the SVG element itself

jameskoster replied:
These are excellent points, but as this PR seeks only to update the design may I suggest we investigate a11y in a dedicated issue?

ciampo replied:
That's a fair point.

Maybe a good interim solution would be to remove `aria-hidden="true"` and change the role to `role="presentation"`. We can then investigate the a11y in a follow-up PR as you suggested

jameskoster committed:
update circle attributes

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+	];
+
+	return (
+		<StyledSpinner className="components-spinner">
```
I believe we may need to keep this hardcoded classname even after switching to Emotion for styles, in order to maintain compat

ciampo commented on packages/components/src/spinner/index.js:
```diff
@@ -1,8 +1,54 @@
+	const strokeDasharray = [
+		Number( CONFIG.spinnerSize ),
+		Number( CONFIG.spinnerSize ) * 10,
+	];
```
Sorry if missed this while skimming through the previous conversations, but is there a reason why we compute dynamically `viewBox`, `radius` and `strokeDasharray`, all in relation to `spinnerSize` ?

With SVG we should be able to define an arbitrary `viewBox` and then scale everything accordingly. We could then use `spinnerSize` in the styles to change `width` and `height` of the `SVG` element itself ([example](https://codepen.io/ciampo/pen/RwLmgPo))

jameskoster replied:
Yeah that would be a lot simpler wouldn't it :) We'll need to keep the computed `stroke-dasharray` though in order for the indicator to scale correctly, so we'll probably need to add a `size` property to the Spinner in the future.

jameskoster committed:
revert entirely to css in js

jameskoster committed:
formatting

ciampo reviewed (commented):
Hey @jameskoster , thank you for your work, and sorry for the late review — I wasn't away of this PR until @sarayourfriend pinged me.

I left a few comments around — would you also be able to add an entry to the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) ? Thank you!

jameskoster committed:
dedicated emotion elements for spinner and indicator

jameskoster committed:
formatting

jameskoster committed:
Number

jameskoster commented on packages/components/src/utils/config-values.js:
```diff
@@ -48,7 +48,7 @@ export default {
 	borderWidthFocus: '1.5px',
 	borderWidthTab: '4px',
-	spinnerSize: '18px',
+	spinnerSize: Number( '18' ),
```
@ciampo let me know if this is correct.

ciampo replied:
You should be able to just use the value `18`, without wrapping it in `Number()`. You could then also update the `stroke-dasharray` attribute, since it doesn't require a runtime `calc` anymore

<details>

<summary>For example (click to expand):</summary>

```diff
diff --git a/packages/components/src/spinner/index.js b/packages/components/src/spinner/index.js
index b256e44a75..f30a852403 100644
--- a/packages/components/src/spinner/index.js
+++ b/packages/components/src/spinner/index.js
@@ -39,8 +39,7 @@ export const SpinnerIndicator = styled.circle`
 	stroke-linecap: round;
 	stroke-width: 1.5px;
 	transform-origin: 50% 50%;
-	stroke-dasharray: ${ CONFIG.spinnerSize },
-		calc( ${ CONFIG.spinnerSize } * 10 );
+	stroke-dasharray: ${ CONFIG.spinnerSize }, ${ CONFIG.spinnerSize * 10 };
 `;
 
 export default function Spinner() {
diff --git a/packages/components/src/utils/config-values.js b/packages/components/src/utils/config-values.js
index 526f85c457..7ba2b9d3dc 100644
--- a/packages/components/src/utils/config-values.js
+++ b/packages/components/src/utils/config-values.js
@@ -48,7 +48,7 @@ export default {
 	borderWidth: '1px',
 	borderWidthFocus: '1.5px',
 	borderWidthTab: '4px',
-	spinnerSize: Number( '18' ),
+	spinnerSize: 18,
 	fontSize: '13px',
 	fontSizeH1: 'calc(2.44 * 13px)',
 	fontSizeH2: 'calc(1.95 * 13px)',
```
</details>

jameskoster replied:
I get an error when I try it this way 🤔 

<img width="876" alt="Screenshot 2022-01-20 at 12 46 21" src="https://user-images.githubusercontent.com/846565/150341268-3ff0db1e-b9da-4c82-be3e-2cbe9fe46060.png">

Thinking about this a little more, do we actually need the `spinnerSize` config value? It doesn't seem to serve much of a purpose right now, especially as all Spinner instances are currently the same size. Might a better approach be to allow a `size` prop on the `Spinner` component? That way it's easy to use different sized spinners ad hoc.

ciampo replied:
> I get an error when I try it this way 🤔

It seems like your IDE thinks that `spinnerSize` is a `string` — did you also update `spinnerSize` in the config (as part of the diff above)? i.e

```suggestion
	spinnerSize: 18,
```

> Thinking about this a little more, do we actually need the spinnerSize config value? It doesn't seem to serve much of a purpose right now, especially as all Spinner instances are currently the same size. Might a better approach be to allow a size prop on the Spinner component? That way it's easy to use different sized spinners ad hoc.

Let's leave it in the config for now, as it matches [the `$spinner-size` variable in the base styles](https://github.com/WordPress/gutenberg/blob/eb8c7690b4cacab028dd74e971a2e000bf5147d6/packages/base-styles/_variables.scss#L58). We have plans later this year to take a closer look at the overall config variables :) 

Re. adding a `size` prop, let's investigate that separately from this PR, maybe while we also address some of the ideas discussed in https://github.com/WordPress/gutenberg/pull/37551#discussion_r787987574

jameskoster replied:
> did you also update spinnerSize in the config (as part of the diff above)? i.e

Derp, I did not :)

> Re. adding a size prop, let's investigate that separately from this PR

Okay! I think we'll need to do _something_, because atm if you use a `Spinner` and resize it with css alone it won't scale correctly:

<img width="234" alt="Screenshot 2022-01-20 at 13 26 18" src="https://user-images.githubusercontent.com/846565/150347718-af58237a-0cc2-4e32-9604-3575d4afdd00.png">

ciampo replied:
Oh, I see. Would that be considered a regression introduced by this PR ? Do we have instances in which the `Spinner` is rendered at a bigger size from its default?

It would still be good to add a Storybook example for this case scenario — (either in this PR or in a separate PR)

ciampo replied:
Alright, I went ahead and had some fun trying out a couple of different approaches — I landed on what seems like a good solution, which avoids using `stroke-dasharray` and therefore scales better when making the spinner bigger (without adding a `size` prop).

These are the steps I went through:

1. Use a `circle` to render the gray "track" (instead of `box-shadow`)
2. Render the indicator (the blue arc) by using the [`path` element](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths#arcs) (instead of `stroke-dasharray`)
3. Clean-up the CSS code by
  - removing the `box-shadow`, `border-radius` and `stroke-dasharray` CSS rules
  - applying the animation only on the indicator arc
  - refactoring some common styles

In order to showcase these changes in Storybook, I went ahead and allowed the `Spinner` component to receive `props` (including a classname). Finally, I added 2 more spinners to the Storybook example:

<img width="343" alt="Screenshot 2022-01-20 at 19 59 02" src="https://user-images.githubusercontent.com/1083581/150404331-d8351268-993b-4753-b567-782e99b1e3ff.png">

<details>

<summary>Click to show code changes</summary>

```diff
diff --git a/packages/components/src/spinner/index.js b/packages/components/src/spinner/index.js
index 3a28c732d8..7ea6a41b91 100644
--- a/packages/components/src/spinner/index.js
+++ b/packages/components/src/spinner/index.js
@@ -1,23 +1,47 @@
+/**
+ * External dependencies
+ */
+import classNames from 'classnames';
+
 /**
  * Internal dependencies
  */
-import { StyledSpinner, SpinnerIndicator } from './styles';
+import { StyledSpinner, SpinnerTrack, SpinnerIndicator } from './styles';
 
-export default function Spinner() {
+/**
+ * @typedef OwnProps
+ *
+ * @property {string} [className] Class name
+ */
+/** @typedef {import('react').ComponentPropsWithoutRef<'svg'> & OwnProps} Props */
+
+/**
+ * @param {Props} props
+ * @return {JSX.Element} Element
+ */
+export default function Spinner( { className, ...props } ) {
 	return (
 		<StyledSpinner
-			className="components-spinner"
+			{ ...props }
+			className={ classNames( 'components-spinner', className ) }
 			viewBox="0 0 100 100"
 			xmlns="http://www.w3.org/2000/svg"
 			role="presentation"
 			focusable="false"
 		>
-			<SpinnerIndicator
-				cx="50%"
-				cy="50%"
+			{ /* Gray circular background */ }
+			<SpinnerTrack
+				cx="50"
+				cy="50"
 				r="50"
 				vectorEffect="non-scaling-stroke"
 			/>
+
+			{ /* Theme-colored arc */ }
+			<SpinnerIndicator
+				d="m 50 0 a 50 50 0 0 1 50 50"
+				vectorEffect="non-scaling-stroke"
+			/>
 		</StyledSpinner>
 	);
 }
diff --git a/packages/components/src/spinner/stories/index.js b/packages/components/src/spinner/stories/index.js
index 12b60f457b..69ab2694a2 100644
--- a/packages/components/src/spinner/stories/index.js
+++ b/packages/components/src/spinner/stories/index.js
@@ -1,10 +1,34 @@
+/**
+ * External dependencies
+ */
+import { css } from '@emotion/react';
+
 /**
  * Internal dependencies
  */
+import { useCx } from '../../utils';
 import Spinner from '../';
 
 export default { title: 'Components/Spinner', component: Spinner };
 
 export const _default = () => {
-	return <Spinner />;
+	const cx = useCx();
+	const mediumSpinner = cx( css`
+		width: 100px;
+		height: 100px;
+	` );
+	const largeSpinner = cx( css`
+		width: 200px;
+		height: 200px;
+	` );
+	return (
+		<>
+			<Spinner />
+			<br />
+			<Spinner className={ mediumSpinner } />
+			<br />
+			<Spinner className={ largeSpinner } />
+			<br />
+		</>
+	);
 };
diff --git a/packages/components/src/spinner/styles.js b/packages/components/src/spinner/styles.js
index 0d2e4c0761..c28c117ce3 100644
--- a/packages/components/src/spinner/styles.js
+++ b/packages/components/src/spinner/styles.js
@@ -2,7 +2,7 @@
  * External dependencies
  */
 import styled from '@emotion/styled';
-import { keyframes } from '@emotion/react';
+import { css, keyframes } from '@emotion/react';
 
 /**
  * Internal dependencies
@@ -10,34 +10,39 @@ import { keyframes } from '@emotion/react';
 import { COLORS, CONFIG } from '../utils';
 
 const spinAnimation = keyframes`
-     from {
-         transform: rotate(0deg);
-     }
-     to {
-         transform: rotate(360deg);
-     }
+	from {
+		transform: rotate(0deg);
+	}
+	to {
+		transform: rotate(360deg);
+	}
  `;
 
 // The Circle can only have a centered stroke so a stacked box-shadow on the svg accomplishes a matching 1.5px border
 export const StyledSpinner = styled.svg`
 	width: ${ CONFIG.spinnerSize }px;
 	height: ${ CONFIG.spinnerSize }px;
-	box-shadow: inset 0 0 0 0.75px ${ COLORS.gray[ 300 ] },
-		0 0 0 0.75px ${ COLORS.gray[ 300 ] };
 	display: block;
-	border-radius: 50%;
 	margin: auto;
 	position: relative;
 	color: var( --wp-admin-theme-color );
 	overflow: visible;
-	animation: 1.4s linear infinite both ${ spinAnimation };
 `;
 
-export const SpinnerIndicator = styled.circle`
+const commonPathProps = css`
 	fill: transparent;
+	stroke-width: 1.5px;
+`;
+
+export const SpinnerTrack = styled.circle`
+	${ commonPathProps };
+	stroke: ${ COLORS.gray[ 300 ] };
+`;
+
+export const SpinnerIndicator = styled.path`
+	${ commonPathProps };
 	stroke: currentColor;
 	stroke-linecap: round;
-	stroke-width: 1.5px;
 	transform-origin: 50% 50%;
-	stroke-dasharray: ${ CONFIG.spinnerSize }, ${ CONFIG.spinnerSize * 10 };
+	animation: 1.4s linear infinite both ${ spinAnimation };
 `;
```
</summary>

I think this approach could solve a couple of problems without introducing much complexity, what do you think?

jameskoster replied:
Oh nice, this looks more comprehensive – thanks so much :) Please feel free to push the changes.

Do you think it's worth including the preset sizes in the component as a part of this PR? If so the only tweak I would suggest would be to use multiples of our grid unit (or spinner-size) values for the medium/large spinners. 

Now that I think of it, it might be a good idea to make the default `spinnerSize` 16px so that it better aligns with the broader grid.

ciampo replied:
> Please feel free to push the changes.

Thank you ! I went ahead and pushed those changes (including a change to the README to remove the out of date screenshot).

> Do you think it's worth including the preset sizes in the component as a part of this PR? If so the only tweak I would suggest would be to use multiples of our grid unit (or spinner-size) values for the medium/large spinners.

I did include the updates to the Storybook examples, if anything to test that the stroke width is not changing as we increase the overall size of the spinner — I also made sure that those storybook examples use multiples of the grid unit.

For now, I wouldn't include any new APIs to the component (e.g. `size`), since the previous one also didn't have it.

> Now that I think of it, it might be a good idea to make the default spinnerSize 16px so that it better aligns with the broader grid.

I haven't made that change yet as I'm not sure if resizing the spinner (from `18px` to `16px`) would have any ripercussions / would be considered breaking change. @jasmussen do you have any opinions on this?

jameskoster replied:
Thanks, this is looking fantastic :) 

I think there's a reasonable case to resize from 18px to 16px. It aligns with our grid unit, and most other icons tend to render at ~16px, e.g. Styles icon:

<img width="170" alt="Screenshot 2022-01-21 at 16 36 14" src="https://user-images.githubusercontent.com/846565/150565004-ea98b958-dee4-435a-99e3-fedaca659c00.png">

Not a strong feeling though, and happy to defer to @jasmussen on this point.

jasmussen replied:
I like it.

ciampo replied:
Done in https://github.com/WordPress/gutenberg/pull/37551/commits/91a861fa30bc0a8ebc957027214d3952c7d79796

Let me know once you've had a look, I think we may be able to merge this PR :)

jameskoster commented:
Thanks @ciampo.

> would you also be able to add an entry to the CHANGELOG ?

I thought this got updated automatically? 🤔

ciampo commented:
> I thought this got updated automatically? 🤔

It is a manual process, at least for the `@wordpress/components` package (can't guarantee for other packages/repos)

jameskoster committed:
split styles into separate file

jameskoster committed:
remove aria-hidden and update role

jameskoster committed:
changelog

jameskoster committed:
update dasharray calc

ciampo reviewed (approved):
Thank you @jameskoster for addressing my feedback! Code changes LGTM — @jasmussen could you also give it a final check in terms of UI/UX in storybook / editor ?

jameskoster commented:
One other thing we'll need to do is update the screenshot in https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/spinner/README.md. Looks like it will need to be uploaded to https://wordpress.org/gutenberg/files/2019/07/spinner.png, I'm not sure how to do that though.

ciampo commented:
> One other thing we'll need to do is update the screenshot in https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/spinner/README.md. Looks like it will need to be uploaded to https://wordpress.org/gutenberg/files/2019/07/spinner.png, I'm not sure how to do that though.

~You can add a new image to this folder (e.g. `packages/components/src/spinner/screenshot.png`) and reference it directly in the README :)~

Edit: Actually, another alternative would be just to remove the preview image — we can assume that folks can visit the [online Storybook showcase](https://wordpress.github.io/gutenberg/) to see the component in action.

jasmussen commented:
Yep, storybook sounds great. It's real, it and it will automatically be up to date, so we won't end up with drift as visuals get tweaked 👌

ciampo committed:
Split track and indicator to separate SVG elements, use a SVG arc for the indicator (instead of stroke-dashoffset)

ciampo committed:
Allow Spinner to accept and forward external props

ciampo committed:
Update storybook example to show different sizes

ciampo committed:
Update README (including the removal of the screenshot)

ciampo committed:
Update spinner size from `18` to `16` px

jameskoster committed:
remove extraneous comment

jameskoster commented:
So close! The only issue outstanding seems to be the link search. For some reason the js styles aren't being applied in that instance:

https://user-images.githubusercontent.com/846565/150759187-06a66eda-5a7b-4de6-96f4-fe18aceafdcc.mp4

Any ideas?

ciampo commented:
I tested on my local machine, using a blank new `wp-env` install, and couldn't reproduce the issue — everything seems fine (apart from a slight vertical misalignment of the spinner, probably due to the new `16px` height):

<img width="1048" alt="Screenshot 2022-01-24 at 11 55 37" src="https://user-images.githubusercontent.com/1083581/150770536-18c9015a-d83f-43ac-bb4a-52b723f730ee.png">

I also tried on [`gutengerg.run`](http://gutenberg.run/37551) and everything looked fine.

Could you share a bit more info on how to reproduce the issue?

---

Edit: ah, I see you're in the template editor, while I was in the post editor. Maybe that makes a difference?

ciampo commented:
Ok, this is weird.

Same WordPress instance.

In the template editor, it looks like the `Spinner` styles (defined through emotion) are not loaded, despite the Emotion-generate classnames being there:

![image](https://user-images.githubusercontent.com/1083581/150773751-301ec752-ec01-4ec3-8cf0-396de696740a.png)

In the post editor, everything looks fine:

![image](https://user-images.githubusercontent.com/1083581/150774113-20b31fcc-9c53-4d74-8f86-03ebbcc00ce7.png)

So, from a first look it seems like there's some issue in loading Emotion's styles in the template editor — @youknowriad and @sarayourfriend , any idea of why that may be happening? Could this be related to [this other conversation](https://github.com/WordPress/gutenberg/pull/35619#discussion_r783270274)?

sarayourfriend commented on packages/components/src/spinner/index.js:
```diff
@@ -1,23 +1,47 @@
+ * @param {Props} props
+ * @return {JSX.Element} Element
+ */
+export default function Spinner( { className, ...props } ) {
```
Typically we forward refs for all components. I guess this is useful for integrating with a library like Reakit where passing refs for important components happens on a regular basis.

Any reason not to foward the ref for Spinner as well?

ciampo replied:
No reason in particular. I guess this PR was initially focused on updating the visual styles of Spinner, and so we didn't focus as much on other aspects (like forwarding refs, rewriting in typescript, using context etcc)

But it's a quick change and it won't hurt to add it to this PR

sarayourfriend replied:
Oh yeah, good point 🙂 Feel free to do in another PR!

If we do forward the refs then we should be able to use `WordPressComponentProps` with it as well.

Might as well do that in a separate PR as it'll require a few other lines to change as well (including calling `useContext` on props).

jameskoster committed:
fix vertical alignment of spinner in link control

ciampo commented on packages/block-editor/src/components/link-control/style.scss:
```diff
@@ -483,16 +483,11 @@ $preview-image-height: 140px;
-		 *  - Horizontally, adjust for the width occupied by the icon buttons,
-		 *    then artificially create spacing that mimics as if the spinner
-		 *    were center-padded to the same width as an icon button.
+		 *  - Input padding top ($grid-unit-15) plus difference between effective input line-height (18px) and spinner height (16px).
+		 *  - Input padding right ($button-size)
```
We should probably translate the calculations explained in the comment in the code too, so that the value stays correct and up to date ?

jameskoster replied:
The only issue there is that the 18px seems to be circumstantial rather than based on any registered value that I can see. 

Thinking about it a little more, we can probably just do something like `top: calc(50% - #{$spinner-size} / 2);` which seems simpler :)

youknowriad commented:
> So, from a first look it seems like there's some issue in loading Emotion's styles in the template editor — @youknowriad and @sarayourfriend , any idea of why that may be happening? Could this be related to this other conversation?

Initially, CSS in JS styles didn't load in iframes by default, we tried solving that (and we did) using the `StyleProvider` in iframe and portals... but lately, it seems that the issues are happening again here and in the related conversation. Unfortunately, this is not really something we know exactly how to fix, it's something in the internals of emotion as our use-case is a bit weird.

sarayourfriend commented:
Yes that does look iframe related.

It seems like iframe styles are broken completely in the editor somehow?

I have a strong suspicion it probably has to do with something being done in the `IFrame` component, as the other isolated parts of the chain for making iframe styles work are unchanged and continue to work in isolation (see the Storybook for `useCx` which renders styles inside an iframe no problem).

I suppose not implementing components in Emotion (or potentially any other CSS-in-JS solution? I believe almost all of them would have this problem, at least in dev mode if not in production as well) is a solution to this, but I believe a lot of components have been implemented in it by this point, so it seems untenable to switch back to sass for all those components.

Solving the iframe style issue should be high priority to avoid having to reimplement things.

jameskoster committed:
Vertical alignment of spinner in link control

jameskoster commented:
How do y'all think we should proceed here? It would be a shame to go back to sass at this point, but I am keen to get this merged.

sarayourfriend commented:
What's odd is that it appears that the spinner works when the site editor is loading, but on subsequent appearances of the component it loses the styles.

I'm prioritizing this today to try to find a fix.

ciampo commented:
> How do y'all think we should proceed here? It would be a shame to go back to sass at this point, but I am keen to get this merged.

I suggest we try to come up with a fix, as it is quite an important issue for the editor and components overall. In case a fix isn't possible or takes too long, we can consider alternative approaches

Edit: just realised Sara also answered, looks like we're aligned :)

sarayourfriend commented:
The issue in this particular case appears to be a problem with how we're using slot/fill.

Take this example:

```
<Iframe>
  <Fill name="popover>
    <Spinner />
  </Fill>
</Iframe>
<Slot name="popover" />
```

The `Spinner` component will receive the Emotion cache context of the `Iframe` component but then be dislocated (in the DOM) outside of the iframe element. This causes the `Spinner` component to render with the incorrect emotion cache and therefore when it renders the CSS for the generated styles are placed in the wrong head element (that of the iframe).

I honestly have zero ideas currently for how to resolve this.

cc @youknowriad

youknowriad commented:
Can we try to maybe add the right `StyleProvider` in `Fill` component or `Slot` (not sure exactly)?

sarayourfriend commented:
Well, I'm stumped again because I tried to reproduce this in Storybook but it seems to just work perfectly:

```js
export const slotFill = () => {
	const redText = css`
		color: red;
	`;
	const blueText = css`
		color: blue;
	`;
	const orangeText = css`
		color: orange;
	`;
	return (
		<SlotFillProvider>
			<IFrame>
				<IFrame>
					<Example args={ [ redText ] }>
						This text is inside an iframe and is red!
					</Example>
					<Fill name="test-slot">
						<Example args={ [ blueText ] }>
							This text is also inside the iframe, but is
							relocated by a slot/fill
						</Example>
					</Fill>
					<Fill name="outside-frame">
						<Example args={ [ orangeText ] }>
							This text is also inside the iframe, but is
							relocated by a slot/fill
						</Example>
					</Fill>
				</IFrame>
				<Slot name="test-slot" />
			</IFrame>
			<Slot name="outside-frame" />
		</SlotFillProvider>
	);
};
```

@youknowriad I tried both adding it to Slot and to Fill and neither had any effect.

I'm going to take a look at everywhere `createPortal` is used and see if there's anything else outstanding.

sarayourfriend commented:
I think I figured it out! No idea how to solve this though 😅 

I believe the problem comes from this pattern:

```js
<Iframe>
  <Spinner />
  <StyleProvider document={ document }>
    <Spinner />
  </StyleProvider>
</Iframe>
```

The first spinner will have the expected styles, but the second, inside the nested StyleProvider, will not.

This is becaue the nested StyleProvider is using a global `document` reference, which will reference the root document of the page rather than the iframe's document.

I was able to reproduce this with slot/fill as well:

```js
<SlotFillProvider>
	<StyleProvider document={ document }>
		<IFrame>
			<IFrame>
				<Example args={ [ redText ] }>
					This text is inside an iframe and is red!
				</Example>
				<Fill name="test-slot">
					<Example args={ [ blueText ] }>
						This text is also inside the iframe, but is
						relocated by a slot/fill
					</Example>
				</Fill>
				<Fill name="outside-frame">
					<Example args={ [ orangeText ] }>
						This text is also inside the iframe, but is
						relocated by a slot/fill
					</Example>
				</Fill>
			</IFrame>
			<StyleProvider document={ document }>
				<Slot name="test-slot" />
			</StyleProvider>
		</IFrame>
		<Slot name="outside-frame" />
	</StyleProvider>
</SlotFillProvider>
```

The text in `test-slot` ends up unstyled.

I have no idea how to fix this though.

sarayourfriend commented:
We have a lot of `StyleProvider` instances that reference the global document and these could be arbitrarily nested inside an `Iframe` component but then transported using slot/fill.

I believe there is no issue with slot/fill, it's just a problem with these StyleProviders getting called with the root document inside the iframe.

ciampo commented:
> I believe there is no issue with slot/fill, it's just a problem with these StyleProviders getting called with the root document inside the iframe

Has this always been an issue (that we only noticed recently), or could it be cause by some recent changes, either on our side or Emotion's side?

sarayourfriend commented:
I'm reluctant to say it's an issue with Emotion as it appears to work entirely as expected in Storybook otherwise.

I haven't been able to solve the problem of the spinner in the site editor. This is what I updated `style-provider/index.js` to, with some debugging code left in to make it easy for anyone coming after me:

```js
//@ts-nocheck

/**
 * WordPress dependencies
 */
import { createContext, useContext } from '@wordpress/element';

/**
 * External dependencies
 */
import { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';
import memoize from 'memize';
import * as uuid from 'uuid';

const uuidCache = new Set();

const memoizedCreateCacheWithContainer = memoize( ( container ) => {
	// emotion only accepts alphabetical and hyphenated keys so we just strip the numbers from the UUID. It _should_ be fine.
	let key = uuid.v4().replace( /[0-9]/g, '' );
	while ( uuidCache.has( key ) ) {
		key = uuid.v4().replace( /[0-9]/g, '' );
	}
	uuidCache.add( key );
	container.dataset.emotionCacheKey = key;
	return createCache( { container, key } );
} );

const heads = [];
window._heads = heads;

/** @type {import('react').Context<HTMLHeadElement>} */
const DocumentHeadContext = createContext( null );

export default function StyleProvider( { children, document } ) {
	const contextHead = useContext( DocumentHeadContext );

	const head = document?.head ?? contextHead ?? null;

	if ( ! head ) return null;

	if ( ! heads.includes( head ) ) heads.push( head );

	if ( head === contextHead ) {
		// If using same head as in context no need to insert a new cache provider
		return children;
	}

	const cache = memoizedCreateCacheWithContainer( head );

	const content = <CacheProvider value={ cache }>{ children }</CacheProvider>;

	return (
		<DocumentHeadContext.Provider value={ head }>
			{ content }
		</DocumentHeadContext.Provider>
	);
}
```

I'm pretty stumped on this, but I might be doing something totally wacky (I haven't thought in "react" in a few months now 🙂) so please review and let me know if there's anything obviously wrong here!

Kudos to @ciampo for the idea to try using React context to avoid using the global document head.

youknowriad commented:
@ciampo @sarayourfriend Thanks for the debugging. I pushed a small commit which I think fixes the issue but let me know if there's anything missing.

The idea is that when you use a `Fill`, the children will be rendered in the document of the `Slot` meaning we need to wrap in another `StyleProvider` otherwise we'll be using the context of the parent of the `Fill` which can be any document/iframe and not necessarily the right one.

ciampo commented:
> Kudos to @ciampo for the idea to try using React context to avoid using the global document head.

It was definitely more of a joint effort with most of the work coming from you 🙌 

> Thanks for the debugging. I pushed a small commit which I think fixes the issue but let me know if there's anything missing.

Thank you @youknowriad! I gave it a quick look at it seems to fix the issue with the `Spinner` in the template editor:


<img width="1792" alt="Screenshot 2022-01-26 at 11 00 35" src="https://user-images.githubusercontent.com/1083581/151143001-8b3bdeac-858c-4f22-8cfa-2725ed166979.png">

I'm going to spend some more time looking into this solution, maybe open a separate PR for it so that the fix doesn't get mixed up / lost with the `Spinner` restyle.

ciampo commented:
Update: I opened a separate PR (#38237) with @youknowriad 's fix where we can discuss the issue separately from this PR (focused on the `Spinner`).

If folks agrees, we can remove Riad's fix from this PR. Once the fix in #38237 gets merged, we'll be able to rebase this PR and merge it too.

jameskoster commented:
> If folks agrees, we can remove Riad's fix from this PR. Once the fix in #38237 gets merged, we'll be able to rebase this PR and merge it too.

Seems sensible to me :)

ciampo reviewed (changes requested):
Cool! In that case, let's wait until #38237 gets merged. We will then rebase and give this PR a final look

ciampo commented:
Alright, #38237 has been merged.

@jameskoster , would you mind rebasing on top of `trunk` to include those changes and solve conflicts? (Feel free to delete Riad's fix on this PR, as it was already included with the merge of #38237).

We can have a final look at the PR and hopefully everything will be in a good shape 🤞

ciampo committed:
Update Storybook example to Controls

ciampo reviewed (approved):
>  would you mind rebasing on top of trunk to include those changes and solve conflicts? (Feel free to delete Riad's fix on this PR, as it was already included with the merge of #38237).

I took the freedom of doing that (hope you don't mind, @jameskoster !)

> We can have a final look at the PR and hopefully everything will be in a good shape 🤞

I gave one final quick look and everything seems to work well — although I'd appreciate if other folks could also give it a final look before merging!

sarayourfriend reviewed (approved):
LGTM! Thanks for all the hard work and patience on this one as we went back and forth on that iframe issue, really appreciate it 🙏

ciampo commented on packages/components/src/spinner/styles.js:
```diff
@@ -0,0 +1,47 @@
+export const StyledSpinner = styled.svg`
+	width: ${ CONFIG.spinnerSize }px;
+	height: ${ CONFIG.spinnerSize }px;
+	display: block;
+	margin: auto;
```
As I gave one last look at the code changes, I just noticed a couple of differences that I'd like to discuss:

- `display`: it used to be `inline-block`, it's currently `block`. I personally like `block` more, but this change could be potentially breaking for some layouts
- `margin`: it used to be `5px 11px 0`, it's not `auto`. This is also a potentially breaking change. I personally believe that a component should not define rules for its own margin, and so my preference would be to either keep things as they are, or to introduce a breaking change by removing the `margin` rule altogether.

What do folks think? (also cc @mirka @jasmussen )

jasmussen replied:
High level I do like the idea of unifying and systematizing how the spinner appears, for example always centered. But it also seems like something that needs a little exploration and guideline in terms of where the spinner is currently used, and might involve updating the places where it's used. That seems to suggest a separate exploratory PR for that, and then in order to land this big improvement, perhaps going back to inline-block. Though I do think we find a better margin, perhaps `$grid-unit-10 $grid-unit-10 0`?

ciampo replied:
I agree with your point.

Probably best thing to do here is to switch back to the values we had before (`display: inline-block` and `margin: 5px 11px 0`), to make sure we don't introduce any breaking changes.

We can then definitely explore what to do separately, as you suggested

jameskoster replied:
I'm unaware of the history around the margin value, but the results are suboptimal in places like the Site Logo and Template Part blocks. `margin: auto;` works better as a default in that regard.

Then again the idea of having no margin on the Spinner at all does seem better on the surface. So I'm happy to leave the margin/display values as-is and revisit separately.

jasmussen replied:
Oh definitely we need that. Just in order to land this one with minimal disruption, we can pull back on those, and then follow up in small PRs, either a separate PR that just re-adds the auto margin entirely, or adds the margin in the contexts where they are needed.

ciampo replied:
Cool, I'll go ahead, push these changes, give it a "final final" look and hopefully merge

ciampo replied:
Merged!

> follow up in small PRs, either a separate PR that just re-adds the auto margin entirely, or adds the margin in the contexts where they are needed.

My preference would be to remove the `margin` rule altogether from `Spinner` and add any necessary `margin` where each instance of `Spinner` is used

jasmussen replied:
Generally agree that the component itself should be unopinionated and subject to its container. But given very few existing containers are likely to be aware of this new responsibility, it might still make sense to have something. We could then consider that a default margin (either a fixed clearance all around, or auto margins), which would then be easily overriden by it's container?

mirka replied:
> My preference would be to remove the `margin` rule altogether from `Spinner` and add any necessary `margin` where each instance of `Spinner` is used

This is my general preference, too. And if the audit shows that it's frequently used as an inline item, I'd consider giving it some padding (instead of margin) so it isn't flush with the bounding box. Basically like an icon component.

pablohoneyhoney commented:
Round of applause!

It really needed an update 😄

jameskoster commented:
Aside from the margin issue above, everything seems to be working well. Thanks for all the help :)

ciampo committed:
Revert `display` and `margin` props to the same values as the previous version of the component

getdave commented:
@jameskoster First up - great work here. This is a big improvement! 🤝 👏 .

I've been looking into improving the communication around loading states in a block I'm working on. I'm curious did you do any research into whether using `speak` from `@wordpress/a11y` would be helpful as part of the component?

I noticed you added `aria-busy` so I'm interested to understand more about this.

Also did anyone from the Core Accessibility team review this PR?

jameskoster commented:
> I noticed you added aria-busy so I'm interested to understand more about this.

I think you are referring to a [different Spinner](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/ui/spinner) :D

In terms of this PR there was [a little discussion](https://github.com/WordPress/gutenberg/pull/37551#discussion_r787979904) around a11y, but the primary focus was to update the visuals.
